File: frontend/src/queries/nodes/InsightQuery/utils/cleanProperties.ts
import {
    AnyPropertyFilter,
    EventPropertyFilter,
    PropertyFilterType,
    PropertyGroupFilter,
    PropertyGroupFilterValue,
    PropertyOperator,
} from '~/types'

/** Cleans properties of insights. These are either a simple list of property filters or a property group filter. The property group filter has
 * a type (AND, OR) and a list of values that are property group filter values, which are either property group filter values or a simple list of
 * property filters.
 */
export const cleanGlobalProperties = (
    properties: Record<string, any> | Record<string, any>[] | undefined
): AnyPropertyFilter[] | PropertyGroupFilter | undefined => {
    if (
        properties == undefined ||
        (Array.isArray(properties) && properties.length == 0) ||
        Object.keys(properties).length == 0
    ) {
        // empty properties
        return undefined
    } else if (isOldStyleProperties(properties)) {
        // old style properties
        properties = transformOldStyleProperties(properties)
        properties = {
            type: 'AND',
            values: [{ type: 'AND', values: properties }],
        }
        return cleanPropertyGroupFilter(properties)
    } else if (Array.isArray(properties)) {
        // list of property filters
        properties = {
            type: 'AND',
            values: [{ type: 'AND', values: properties }],
        }
        return cleanPropertyGroupFilter(properties)
    } else if (
        (properties['type'] === 'AND' || properties['type'] === 'OR') &&
        !properties['values'].some((property: any) => property['type'] === 'AND' || property['type'] === 'OR')
    ) {
        // property group filter value
        properties = {
            type: 'AND',
            values: [properties],
        }
        return cleanPropertyGroupFilter(properties)
    }
    // property group filter
    return cleanPropertyGroupFilter(properties)
}

/** Cleans properties of entities i.e. event and action nodes. These are a simple list of property filters. */
export const cleanEntityProperties = (
    properties: Record<string, any> | Record<string, any>[] | undefined
): AnyPropertyFilter[] | undefined => {
    if (
        properties == undefined ||
        (Array.isArray(properties) && properties.length == 0) ||
        Object.keys(properties).length == 0
    ) {
        // empty properties
        return undefined
    } else if (isOldStyleProperties(properties)) {
        // old style properties
        return transformOldStyleProperties(properties)
    } else if (Array.isArray(properties)) {
        // list of property filters
        return properties.map(cleanProperty)
    } else if (
        (properties['type'] === 'AND' || properties['type'] === 'OR') &&
        !properties['values'].some((property: any) => property['type'] === 'AND' || property['type'] === 'OR')
    ) {
        // property group filter value
        return properties.values.map(cleanProperty)
    }
    throw new Error('Unexpected format of entity properties.')
}

const cleanPropertyGroupFilter = (properties: Record<string, any>): PropertyGroupFilter => {
    properties['values'] = cleanPropertyGroupFilterValues(properties.values)
    return properties as PropertyGroupFilter
}

const cleanPropertyGroupFilterValues = (
    properties: (AnyPropertyFilter | PropertyGroupFilterValue)[]
): (AnyPropertyFilter | PropertyGroupFilterValue)[] => {
    return properties.map(cleanPropertyGroupFilterValue)
}

const cleanPropertyGroupFilterValue = (
    property: AnyPropertyFilter | PropertyGroupFilterValue
): AnyPropertyFilter | PropertyGroupFilterValue => {
    if (property['type'] == 'AND' || property['type'] == 'OR') {
        // property group filter value
        property['values'] = cleanPropertyGroupFilterValues(property['values'] as PropertyGroupFilterValue[])
        return property
    }
    // property filter
    return cleanProperty(property)
}

const cleanProperty = (property: Record<string, any>): AnyPropertyFilter => {
    if (Object.keys(property).length === 0) {
        return { type: PropertyFilterType.HogQL, key: 'true' }
    }

    // remove invalid properties without type
    if (property['type'] === undefined) {
        return { type: PropertyFilterType.HogQL, key: 'true' }
    }

    // fix type typo
    if (property['type'] === 'events') {
        property['type'] = 'event'
    }

    // fix value key typo
    if (property['values'] !== undefined && property['value'] === undefined) {
        property['value'] = property['values']
        delete property['values']
    }

    // convert precalculated and static cohorts to cohorts
    if (['precalculated-cohort', 'static-cohort'].includes(property['type'])) {
        property['type'] = 'cohort'
    }

    // fix invalid property key for cohorts
    if (property['type'] === 'cohort' && property['key'] !== 'id') {
        property['key'] = 'id'
    }

    // set a default operator for properties that support it, but don't have an operator set
    if (isPropertyWithOperator(property) && property['operator'] === undefined) {
        property['operator'] = 'exact'
    }

    // remove the operator for properties that don't support it, but have it set
    if (!isPropertyWithOperator(property) && property['operator'] !== undefined) {
        delete property['operator']
    }

    // convert `negation` for cohorts
    if (property['type'] === 'cohort' && property['negation'] !== undefined) {
        if (property['operator'] === PropertyOperator.Exact && property['negation']) {
            property['operator'] = PropertyOperator.NotIn
        }
        delete property['negation']
    }

    // remove none from values
    if (Array.isArray(property['value'])) {
        property['value'] = property['value'].filter((x) => x !== null)
    }

    // remove keys without concrete value
    Object.keys(property).forEach((key) => {
        if (property[key] === undefined) {
            delete property[key]
        }
    })

    return property
}

const isPropertyWithOperator = (property: Record<string, any>): boolean => {
    return !['hogql'].includes(property['type'])
}

// old style dict properties e.g. {"utm_medium__icontains": "email"}
const isOldStyleProperties = (properties: Record<string, any> | Record<string, any>[]): boolean => {
    return (
        !Array.isArray(properties) && Object.keys(properties).length === 1 && !['AND', 'OR'].includes(properties.type)
    )
}

const transformOldStyleProperties = (
    properties: Record<string, any> | Record<string, any>[]
): EventPropertyFilter[] => {
    const key = Object.keys(properties)[0]
    const value = Object.values(properties)[0]
    const keySplit = key.split('__')

    return [
        {
            key: keySplit[0],
            value: value,
            operator: keySplit.length > 1 ? (keySplit[1] as PropertyOperator) : PropertyOperator.Exact,
            type: PropertyFilterType.Event,
        },
    ]
}

File: posthog/hogql_queries/legacy_compatibility/clean_properties.py
from posthog.schema import PropertyOperator


def clean_global_properties(properties: dict | list[dict] | None):
    if properties is None or len(properties) == 0:
        # empty properties
        return None
    elif is_old_style_properties(properties):
        # old style properties
        properties = transform_old_style_properties(properties)
        properties = {
            "type": "AND",
            "values": [{"type": "AND", "values": properties}],
        }
        return clean_property_group_filter(properties)
    elif isinstance(properties, list):
        # list of property filters
        properties = {
            "type": "AND",
            "values": [{"type": "AND", "values": properties}],
        }
        return clean_property_group_filter(properties)
    elif (
        isinstance(properties, dict)
        and properties.get("type") in ["AND", "OR"]
        and not any(property.get("type") in ["AND", "OR"] for property in properties["values"])
    ):
        # property group filter value
        properties = {
            "type": "AND",
            "values": [properties],
        }
        return clean_property_group_filter(properties)
    else:
        # property group filter
        return clean_property_group_filter(properties)


def clean_entity_properties(properties: list[dict] | dict | None):
    if properties is None or len(properties) == 0:
        # empty properties
        return None
    elif is_old_style_properties(properties):
        # old style properties
        return transform_old_style_properties(properties)
    elif isinstance(properties, list):
        # list of property filters
        return list(map(clean_property, properties))
    elif (
        isinstance(properties, dict)
        and properties.get("type") in ["AND", "OR"]
        and not any(property.get("type") in ["AND", "OR"] for property in properties["values"])
    ):
        # property group filter value
        return list(map(clean_property, properties["values"]))
    else:
        raise ValueError("Unexpected format of entity properties.")


def clean_property_group_filter(properties: dict):
    properties["values"] = clean_property_group_filter_values(properties["values"])
    return properties


def clean_property_group_filter_values(properties: list[dict]):
    cleaned = [clean_property_group_filter_value(property) for property in properties if property]
    return cleaned


def clean_property_group_filter_value(property: dict):
    if property.get("type") in ["AND", "OR"]:
        # property group filter value
        property["values"] = clean_property_group_filter_values(property["values"])
        return property
    else:
        # property filter
        return clean_property(property)


def clean_property(property: dict):
    cleaned_property = {**property}

    # fix type typo
    if cleaned_property.get("type") == "events":
        cleaned_property["type"] = "event"

    # fix value key typo
    if cleaned_property.get("values") is not None and cleaned_property.get("value") is None:
        cleaned_property["value"] = cleaned_property.pop("values")

    # convert precalculated and static cohorts to cohorts
    if cleaned_property.get("type") in ("precalculated-cohort", "static-cohort"):
        cleaned_property["type"] = "cohort"

    # fix invalid property key for cohorts
    if cleaned_property.get("type") == "cohort":
        if cleaned_property.get("key") != "id":
            cleaned_property["key"] = "id"
        if cleaned_property.get("operator") is None:
            cleaned_property["operator"] = (
                PropertyOperator.NOT_IN.value if cleaned_property.get("negation", False) else PropertyOperator.IN_.value
            )
        if "negation" in cleaned_property:
            del cleaned_property["negation"]

    # set a default operator for properties that support it, but don't have an operator set
    if is_property_with_operator(cleaned_property) and cleaned_property.get("operator") is None:
        cleaned_property["operator"] = "exact"

    # remove the operator for properties that don't support it, but have it set
    if not is_property_with_operator(cleaned_property) and cleaned_property.get("operator") is not None:
        del cleaned_property["operator"]

    # remove none from values
    if isinstance(cleaned_property.get("value"), list):
        cleaned_property["value"] = list(filter(lambda x: x is not None, cleaned_property.get("value")))  # type: ignore

    # remove keys without concrete value
    cleaned_property = {key: value for key, value in cleaned_property.items() if value is not None}

    return cleaned_property


def is_property_with_operator(property: dict):
    return property.get("type") not in ("hogql",)


# old style dict properties e.g. {"utm_medium__icontains": "email"}
def is_old_style_properties(properties):
    return isinstance(properties, dict) and len(properties) == 1 and properties.get("type") not in ("AND", "OR")


def transform_old_style_properties(properties):
    key = next(iter(properties.keys()))
    value = next(iter(properties.values()))
    key_split = key.split("__")
    return [
        {
            "key": key_split[0],
            "value": value,
            "operator": key_split[1] if len(key_split) > 1 else "exact",
            "type": "event",
        }
    ]
